src/api-service/__app__/onefuzzlib/azure/nsg.py

```diff
@@ -0,0 +1,484 @@
+            logging.debug(
+                "create NSG had conflicts with concurrent request, ignoring %s", err
+            )
+            return None
+        return Error(
+            code=ErrorCode.UNABLE_TO_CREATE,
+            errors=["Unable to update nsg %s due to %s" % (nsg.name, err)],
+        )
+    return None
+
+
+# Return True if NSG is created using OneFuzz naming convention.
+# Therefore NSG belongs to OneFuzz.
+def ok_to_delete(active_regions: Set[Region], nsg_region: str, nsg_name: str) -> bool:
+    return nsg_region not in active_regions and nsg_region == nsg_name
+
+
+def is_onefuzz_nsg(nsg_region: str, nsg_name: str) -> bool:
+    return nsg_region == nsg_name
+
+
+# Returns True if deletion completed (thus resource not found) or successfully started.
+# Returns False if failed to start deletion.
+def start_delete_nsg(name: str) -> bool:
+    # NSG can be only deleted if no other resource is associated with it
+    resource_group = get_base_resource_group()
+
+    logging.info("deleting nsg: %s %s", resource_group, name)
+    network_client = get_network_client()
+
+    try:
+        network_client.network_security_groups.begin_delete(resource_group, name)
+        return True
+    except HttpResponseError as err:
+        err_str = str(err)
+        if (
+            "cannot be deleted because it is in use by the following resources"
+        ) in err_str:
+            return False
+    except ResourceNotFoundError:
+        return True
+
+    return False
+
+
+def set_allowed(name: str, sources: NetworkSecurityGroupConfig) -> Union[None, Error]:
+    resource_group = get_base_resource_group()
+    nsg = get_nsg(name)
+    if not nsg:
+        return Error(
+            code=ErrorCode.UNABLE_TO_FIND,
+            errors=["cannot update nsg rules. nsg %s not found" % name],
+        )
+
+    logging.info(
+        "setting allowed incoming connection sources for nsg: %s %s",
+        resource_group,
+        name,
+    )
+    all_sources = sources.allowed_ips + sources.allowed_service_tags
+    security_rules = []
+    # NSG security rule priority range defined here:
+    # https://docs.microsoft.com/en-us/azure/virtual-network/network-security-groups-overview
+    min_priority = 100
+    # NSG rules per NSG limits:
+    # https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/azure-subscription-service-limits?toc=/azure/virtual-network/toc.json#networking-limits
+    max_rule_count = 1000
+    if len(all_sources) > max_rule_count:
+        return Error(
+            code=ErrorCode.INVALID_REQUEST,
+            errors=[
+                "too many rules provided %d. Max allowed: %d"
+                % ((len(all_sources)), max_rule_count),
+            ],
+        )
+
+    priority = min_priority
+    for src in all_sources:
+        security_rules.append(
+            SecurityRule(
+                name="Allow" + str(priority),
+                protocol="*",
+                source_port_range="*",
```

Comment:
Should we be restricting ports, too?

Reply:
is it a requirement at this point ? 

If request is coming from trusted IP  - might want to ssh/RDP to a debug VM, and have a debugger listening on a port of their choosing, no ?

Reply:
Just thinking aloud. WDYT?

Reply:
my thinking that currently we just want to control IPs only, since IP is trusted and OneFuzz is a "dev tool" trusted people  should be able to do what they think is needed to be done.

Reply:
I think that since these NSGs apply to ephemeral proxy VMs, and we only support public key auth with short-lived keys, this is a solid starting point, and is a strict improvement on the status quo. We can investigate further restriction later (will open an issue for it).

Reply:
Captured in: #1436.